File: implementations/python/recursion_and_backtracking.py
"""
Recursion and Backtracking implementation in Python.

Provides:
1. N-Queens Solver with diagonal conflict tracking.
2. Power set / Subset generation (with and without duplicate handling).
3. Permutation generation.
4. Combination Sum with pruning.
5. Comprehensive unit tests.
"""

import unittest
from typing import List


class NQueens:
    """N-Queens solver using backtracking with conflict tracking."""

    def __init__(self, n: int):
        self.n = n

    def solve(self) -> List[List[str]]:
        if self.n <= 0:
            return []

        solutions: List[List[str]] = []
        col_placement = [-1] * self.n
        used_cols = set()
        used_diag1 = set()  # r + c
        used_diag2 = set()  # r - c

        def backtrack(row: int) -> None:
            if row == self.n:
                board = []
                for r in range(self.n):
                    c = col_placement[r]
                    row_str = '.' * c + 'Q' + '.' * (self.n - c - 1)
                    board.append(row_str)
                solutions.append(board)
                return

            for c in range(self.n):
                d1 = row + c
                d2 = row - c
                if c in used_cols or d1 in used_diag1 or d2 in used_diag2:
                    continue  # Pruned

                # Choose
                col_placement[row] = c
                used_cols.add(c)
                used_diag1.add(d1)
                used_diag2.add(d2)

                # Explore
                backtrack(row + 1)

                # Unchoose
                used_cols.remove(c)
                used_diag1.remove(d1)
                used_diag2.remove(d2)

        backtrack(0)
        return solutions

    def count_solutions(self) -> int:
        return len(self.solve())


def generate_subsets(nums: List[int]) -> List[List[int]]:
    """Generate power set using choose-explore-unchoose backtracking."""
    result: List[List[int]] = []
    curr: List[int] = []

    def backtrack(idx: int) -> None:
        if idx == len(nums):
            result.append(list(curr))
            return
        # Branch 1: exclude
        backtrack(idx + 1)
        # Branch 2: include
        curr.append(nums[idx])
        backtrack(idx + 1)
        curr.pop()

    backtrack(0)
    return result


def generate_subsets_unique(nums: List[int]) -> List[List[int]]:
    """Generate unique subsets when input contains duplicates (Subsets II)."""
    sorted_nums = sorted(nums)
    result: List[List[int]] = []
    curr: List[int] = []

    def backtrack(start: int) -> None:
        result.append(list(curr))
        for i in range(start, len(sorted_nums)):
            if i > start and sorted_nums[i] == sorted_nums[i - 1]:
                continue
            curr.append(sorted_nums[i])
            backtrack(i + 1)
            curr.pop()

    backtrack(0)
    return result


def generate_permutations(nums: List[int]) -> List[List[int]]:
    """Generate all permutations using backtracking and swapping."""
    result: List[List[int]] = []
    arr = list(nums)

    def backtrack(start: int) -> None:
        if start == len(arr):
            result.append(list(arr))
            return
        for i in range(start, len(arr)):
            arr[start], arr[i] = arr[i], arr[start]
            backtrack(start + 1)
            arr[start], arr[i] = arr[i], arr[start]

    backtrack(0)
    return result


def combination_sum(candidates: List[int], target: int) -> List[List[int]]:
    """Find all unique combinations summing to target with candidate reuse."""
    candidates = sorted(candidates)
    result: List[List[int]] = []
    curr: List[int] = []

    def backtrack(start: int, remain: int) -> None:
        if remain == 0:
            result.append(list(curr))
            return
        for i in range(start, len(candidates)):
            if candidates[i] > remain:
                break  # Prune
            curr.append(candidates[i])
            backtrack(i, remain - candidates[i])
            curr.pop()

    backtrack(0, target)
    return result


class TestRecursionAndBacktracking(unittest.TestCase):
    def test_n_queens(self):
        self.assertEqual(NQueens(1).count_solutions(), 1)
        self.assertEqual(NQueens(2).count_solutions(), 0)
        self.assertEqual(NQueens(3).count_solutions(), 0)
        self.assertEqual(NQueens(4).count_solutions(), 2)
        sols4 = NQueens(4).solve()
        self.assertEqual(len(sols4), 2)
        self.assertEqual(len(sols4[0]), 4)
        self.assertEqual(NQueens(8).count_solutions(), 92)

    def test_subsets(self):
        subs = generate_subsets([1, 2, 3])
        self.assertEqual(len(subs), 8)

        uniq_subs = generate_subsets_unique([1, 2, 2])
        self.assertEqual(len(uniq_subs), 6)

    def test_permutations(self):
        perms = generate_permutations([1, 2, 3])
        self.assertEqual(len(perms), 6)
        self.assertIn([1, 2, 3], perms)
        self.assertIn([3, 2, 1], perms)

        single = generate_permutations([9])
        self.assertEqual(single, [[9]])

    def test_combination_sum(self):
        res = combination_sum([2, 3, 6, 7], 7)
        self.assertEqual(res, [[2, 2, 3], [7]])
        self.assertEqual(combination_sum([2], 1), [])


if __name__ == '__main__':
    unittest.main()

File: implementations/cpp/recursion_and_backtracking.cpp
#include <cassert>
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

namespace dsa {

/**
 * @brief N-Queens solver using backtracking with O(1) conflict checking.
 */
class NQueens {
private:
    int n_;
    std::vector<std::vector<std::string>> solutions_;
    std::vector<int> col_placement_; // col_placement_[row] = col
    std::vector<bool> used_cols_;
    std::vector<bool> used_diag1_;   // r + c
    std::vector<bool> used_diag2_;   // r - c + n - 1

    void backtrack(int row) {
        if (row == n_) {
            // Found a valid solution, construct board
            std::vector<std::string> board(n_, std::string(n_, '.'));
            for (int r = 0; r < n_; ++r) {
                board[r][col_placement_[r]] = 'Q';
            }
            solutions_.push_back(std::move(board));
            return;
        }

        for (int c = 0; c < n_; ++c) {
            int d1 = row + c;
            int d2 = row - c + n_ - 1;

            if (used_cols_[c] || used_diag1_[d1] || used_diag2_[d2]) {
                continue; // Pruned
            }

            // Choose
            col_placement_[row] = c;
            used_cols_[c] = true;
            used_diag1_[d1] = true;
            used_diag2_[d2] = true;

            // Explore
            backtrack(row + 1);

            // Unchoose
            used_cols_[c] = false;
            used_diag1_[d1] = false;
            used_diag2_[d2] = false;
        }
    }

public:
    explicit NQueens(int n)
        : n_(n), col_placement_(n, -1),
          used_cols_(n, false), used_diag1_(2 * n, false), used_diag2_(2 * n, false) {}

    std::vector<std::vector<std::string>> solve() {
        solutions_.clear();
        if (n_ <= 0) return solutions_;
        backtrack(0);
        return solutions_;
    }

    int count_solutions() {
        return static_cast<int>(solve().size());
    }
};

/**
 * @brief Generate all subsets (power set) of a vector using backtracking.
 */
inline void subsets_helper(const std::vector<int>& nums, size_t idx,
                           std::vector<int>& current,
                           std::vector<std::vector<int>>& result) {
    if (idx == nums.size()) {
        result.push_back(current);
        return;
    }

    // Branch 1: Exclude nums[idx]
    subsets_helper(nums, idx + 1, current, result);

    // Branch 2: Include nums[idx]
    current.push_back(nums[idx]);
    subsets_helper(nums, idx + 1, current, result);
    current.pop_back(); // Unchoose
}

inline std::vector<std::vector<int>> generate_subsets(const std::vector<int>& nums) {
    std::vector<std::vector<int>> result;
    std::vector<int> current;
    subsets_helper(nums, 0, current, result);
    return result;
}

/**
 * @brief Generate all subsets with duplicates handled (Subsets II).
 */
inline void subsets_with_dup_helper(const std::vector<int>& nums, size_t start,
                                    std::vector<int>& current,
                                    std::vector<std::vector<int>>& result) {
    result.push_back(current);
    for (size_t i = start; i < nums.size(); ++i) {
        if (i > start && nums[i] == nums[i - 1]) {
            continue; // Skip duplicate branch
        }
        current.push_back(nums[i]);
        subsets_with_dup_helper(nums, i + 1, current, result);
        current.pop_back();
    }
}

inline std::vector<std::vector<int>> generate_subsets_unique(std::vector<int> nums) {
    std::sort(nums.begin(), nums.end());
    std::vector<std::vector<int>> result;
    std::vector<int> current;
    subsets_with_dup_helper(nums, 0, current, result);
    return result;
}

/**
 * @brief Generate all permutations using backtracking and in-place swapping.
 */
inline void permutations_helper(std::vector<int>& nums, size_t start,
                                std::vector<std::vector<int>>& result) {
    if (start == nums.size()) {
        result.push_back(nums);
        return;
    }
    for (size_t i = start; i < nums.size(); ++i) {
        std::swap(nums[start], nums[i]);       // Choose
        permutations_helper(nums, start + 1, result); // Explore
        std::swap(nums[start], nums[i]);       // Unchoose
    }
}

inline std::vector<std::vector<int>> generate_permutations(std::vector<int> nums) {
    std::vector<std::vector<int>> result;
    permutations_helper(nums, 0, result);
    return result;
}

/**
 * @brief Combination Sum: Find all unique combinations where candidates sum to target.
 * Candidates can be reused unlimited times.
 */
inline void combination_sum_helper(const std::vector<int>& candidates, int target,
                                   size_t start, std::vector<int>& current,
                                   std::vector<std::vector<int>>& result) {
    if (target == 0) {
        result.push_back(current);
        return;
    }
    for (size_t i = start; i < candidates.size(); ++i) {
        if (candidates[i] > target) {
            break; // Pruned because array is sorted
        }
        current.push_back(candidates[i]);
        combination_sum_helper(candidates, target - candidates[i], i, current, result);
        current.pop_back();
    }
}

inline std::vector<std::vector<int>> combination_sum(std::vector<int> candidates, int target) {
    std::sort(candidates.begin(), candidates.end());
    std::vector<std::vector<int>> result;
    std::vector<int> current;
    combination_sum_helper(candidates, target, 0, current, result);
    return result;
}

} // namespace dsa

int main() {
    using namespace dsa;

    std::cout << "[RUNNING] Recursion and Backtracking C++17 Verification..." << std::endl;

    // 1. N-Queens Verification
    {
        NQueens q1(1);
        assert(q1.count_solutions() == 1);

        NQueens q2(2);
        assert(q2.count_solutions() == 0);

        NQueens q3(3);
        assert(q3.count_solutions() == 0);

        NQueens q4(4);
        assert(q4.count_solutions() == 2);
        auto sols4 = q4.solve();
        assert(sols4.size() == 2);
        assert(sols4[0].size() == 4);

        NQueens q8(8);
        assert(q8.count_solutions() == 92);
    }

    // 2. Subset Generation
    {
        std::vector<int> nums = {1, 2, 3};
        auto subsets = generate_subsets(nums);
        assert(subsets.size() == 8);

        std::vector<int> dups = {1, 2, 2};
        auto unique_subsets = generate_subsets_unique(dups);
        assert(unique_subsets.size() == 6); // [], [1], [1,2], [1,2,2], [2], [2,2]
    }

    // 3. Permutation Generation
    {
        std::vector<int> nums = {1, 2, 3};
        auto perms = generate_permutations(nums);
        assert(perms.size() == 6);

        std::vector<int> single = {5};
        auto single_perms = generate_permutations(single);
        assert(single_perms.size() == 1);
        assert((single_perms[0] == std::vector<int>{5}));
    }

    // 4. Combination Sum with Pruning
    {
        std::vector<int> cands = {2, 3, 6, 7};
        auto res = combination_sum(cands, 7);
        // [2, 2, 3] and [7]
        assert(res.size() == 2);
        assert((res[0] == std::vector<int>{2, 2, 3}));
        assert((res[1] == std::vector<int>{7}));

        auto res_empty = combination_sum({2}, 1);
        assert(res_empty.empty());
    }

    std::cout << "[PASSED] Recursion and Backtracking C++17 All Tests Passed!" << std::endl;
    return 0;
}
